h3/src/client.rs

```diff
@@ -420,22 +429,18 @@ where
                     )))
                 }
                 Err(e) => {
-                    let connection_error = self
-                        .inner
-                        .shared
-                        .read("poll_close error read")
-                        .error
-                        .as_ref()
-                        .cloned();
-
-                    match connection_error {
-                        Some(e) if e.is_closed() => return Poll::Ready(Ok(())),
-                        Some(e) => return Poll::Ready(Err(e)),
+                    let connection_error = self.inner.shared.read("poll_close").error.clone();
+                    let connection_error = match connection_error {
+                        Some(e) => e,
                         None => {
-                            self.inner.shared.write("poll_close error").error = e.clone().into();
-                            return Poll::Ready(Err(e));
+                            self.inner.shared.write("poll_close error").error = Some(e.clone());
+                            e
                         }
+                    };
+                    if connection_error.is_closed() {
+                        return Poll::Ready(Ok(()));
                     }
+                    return Poll::Ready(Err(connection_error));
```

Comment:
I think: This fixes a preexisting bug, which wasn't discovered due to the lack of test coverage of client shutdown (cases where the server closes the connection without error). Basically, if the server closes, the first time the resulting error on the client is seen, the old code does the match on the previous error in `self.inner.shared.error` (which is `None`), thus skipping the arm that leads to `Ok(())`.

The only way I could make this make sense to me was that the intention of this piece of code is to match on the first error, but got the order of doing things confused.
So I did the equivalent to `let connection_error = self.inner.shared.write("poll_close").error.get_or_insert(e)` but without taking the write lock if not necessary.

Did I misunderstand the original intent here?

Reply:
I do not know what the original intention was. But the change makes sense to me.

Reply:
> the lack of test coverage of client shutdown

Thanks for helping to add some, more tests is love! <3